052-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,171 @@
+1. The jmx_exporter Java agent supports complex metrics mapping rules. These rules allow renaming metrics and it can make it hard to investigate issues as metrics could have different names. This prevents using metrics like an API, whether this is to build grafana dashboards or for the operator to rely on metrics.
+1. Due to the complex rules, it performs badly when there’s a very large number of metrics due to a lot of topic/partitions.
+1. Recently jmx_exporter was impacted by CVEs in its dependencies (CVE-2022-38752, CVE-2022-25857).
+
+
+## Proposal
+
+The proposal is to build metrics reporters that directly exposes metrics via an HTTP endpoint in the Prometheus format. This will be a new project/repository under the Strimzi organization.
+
+![Proposal](images/052-proposal.png)
+
+Today to enable jmx_exporter, users use:
+```
+metricsConfig:
+  type: jmxPrometheusExporter
+  valueFrom:
+    configMapKeyRef:
+      name: kafka-metrics
+      key: kafka-metrics-config.yml
+```
+
+To enable metric reporters, they will instead use this for example:
+```
+metricsConfig:
+  config:
+    allowlist: ".*"
+    port: 8080
+```
+
+Reporters will expose the following configurations:
+- `prometheus.metrics.reporter.port`: The HTTP port to expose the metrics. Default: `8080`
+- `prometheus.metrics.reporter.allowlist`: A comma separated list of regex patterns to specify the metrics to collect. Default: `.*`. The patterns must match the metric names emitted by the reporter.
+
+The reporter will also export JVM metrics similar to the ones exported by jmx_exporter. These are provided by the [hotspot exports](io.prometheus.client.hotspot) from the Prometheus Java client.
+
+This proposal ignores supporting HTTPS as today Strimzi does not allow configuring it with jmx_exporter. Note that jmx_exporter 0.19.0 added support for HTTPS. If needed we can add it now or later to the reporter.
+
+In Kafka, metric reporters are configured per Kafka client or broker. This makes it hard to use reporters in client-side applications that use multiple clients such as Connect, Streams but also Strimzi components like the HTTP bridge. At the moment, for these components it's preferable to keep the current metrics pipeline with JMX and jmx_exporter.
+
+Today ZooKeeper also uses jmx_exporter to expose its metrics. ZooKeeper is going away and will be replaced by KRaft controllers which also support Kafka metrics reporters.
+
+For these two reasons, I propose only using reporters for brokers/controllers and only using them in KRaft mode.
+
+
+## Affected/not affected projects
+
+This affects `strimzi-kafka-operator `. Since ZooKeeper does not support metrics reporters, we should keep the current mechanism using jmx_exporter while we support ZooKeeper deployments.
```

Comment:
As you said above, its not just ZooKeeper but also Connect / Mirror Maker. Maybe adding it here will make it more clear and obvious as I requested above 🤔